File: py/minsk/analysis/binding/statements/variable.py
from dataclasses import dataclass

from minsk.analysis.binding.expression import BoundExpression
from minsk.analysis.binding.kind import BoundNodeKind
from minsk.analysis.binding.statement import BoundStatement
from minsk.analysis.variable import VariableSymbol


@dataclass(frozen=True)
class BoundVariableDeclaration(BoundStatement):
    variable: VariableSymbol
    initializer: BoundExpression

    @property
    def kind(self) -> BoundNodeKind:
        return BoundNodeKind.VariableDeclaration

File: cpp/minsk/source/analysis/binding/nodes/statements/variable.cpp
#include "minsk/analysis/binding/nodes/statements/variable.hpp"
#include "minsk/analysis/binding/kind.hpp"

minsk::analysis::binding::bound_variable_declaration::
    bound_variable_declaration(variable_symbol &&variable,
                               std::unique_ptr<bound_expression> initializer)
    : m_variable(std::move(variable)), m_initializer(std::move(initializer)) {}

minsk::analysis::binding::bound_node_kind
minsk::analysis::binding::bound_variable_declaration::kind() const {
  return bound_node_kind::variable_declaration;
}

const minsk::analysis::variable_symbol &
minsk::analysis::binding::bound_variable_declaration::variable() const {
  return m_variable;
}

const minsk::analysis::binding::bound_expression *
minsk::analysis::binding::bound_variable_declaration::initializer() const {
  return m_initializer.get();
}
